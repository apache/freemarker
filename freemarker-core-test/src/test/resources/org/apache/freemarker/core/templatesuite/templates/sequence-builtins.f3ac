<#--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->
<@noOutput>
<#setting locale="en_US">
<#setting numberFormat="0.#########">

<#assign ls = []?sort>
<#list ls as i>
- ${i}
</#list>
<@assertEquals expected=0 actual=ls?size />
<@assertEquals expected=3 actual=set?size />
</@noOutput>
Sorting strings:
----------------

String order:
<#assign ls = ["whale", "Barbara", "zeppelin", "aardvark", "beetroot"]?sort>
<#list ls as i>
- ${i}
</#list>

First: ${ls?first}
Last: ${ls?last}
Size ${ls?size}

Numerical order:
<#assign ls = [123?byte, 543, -324, -34?float, 0.11, 0, 111?int, 0.1?double, 1, 5]?sort>
<#list ls as i>
- ${i}
</#list>

First: ${ls?first}
Last: ${ls?last}
Size ${ls?size}

Date/time order:
<#assign x = [
        '08:05'?time('HH:mm'),
        '18:00'?time('HH:mm'),
        '06:05'?time('HH:mm'),
        '08:15'?time('HH:mm')]>
<#list x?sort as i>
- ${i?string('HH:mm')}
</#list>

Boolean order:
<#assign x = [
        true,
        false,
        false,
        true]>
<#list x?sort as i>
- ${i?string}
</#list>


Sorting hashes:
---------------

<#assign ls = [
  {"name":"whale", "weight":2000?short},
  {"name":"Barbara", "weight":53},
  {"name":"zeppelin", "weight":-200?float},
  {"name":"aardvark", "weight":30?long},
  {"name":"beetroot", "weight":0.3}
]>
Order by name:
<#assign ls = ls?sortBy("name")>
<#list ls as i>
- ${i.name}: ${i.weight}
</#list>

Order by weight:
<#assign ls = ls?sortBy("weight")>
<#list ls as i>
- ${i.name}: ${i.weight}
</#list>

Order by a.x.v:
<#assign x = [
        {"a": {"x": {"v": "qweqw", "w": "asd"}, "y": '1998-02-20'?date('yyyy-MM-dd')}},
        {"a": {"x": {"v": "aqweqw", "w": "asd"}, "y": '1999-01-20'?date('yyyy-MM-dd')}},
        {"a": {"x": {"v": "dfgdf", "w": "asd"}, "y": '1999-04-20'?date('yyyy-MM-dd')}},
        {"a": {"x": {"v": "utyu", "w": "asd"}, "y": '1999-04-19'?date('yyyy-MM-dd')}}]>
<#list x?sortBy(['a', 'x', 'v']) as i>
- ${i.a.x.v}
</#list>

Order by a.y, which is a date:
<#list x?sortBy(['a', 'y']) as i>
- ${i.a.y?string('yyyy-MM-dd')}
</#list>

Reverse:
--------

Order by weight desc:
<#assign ls = ls?reverse>
<#list ls as i>
- ${i.name}: ${i.weight}
</#list>

Order by weight desc desc:
<#assign ls = ls?reverse>
<#list ls as i>
- ${i.name}: ${i.weight}
</#list>

Order by weight desc desc desc:
<#assign ls = ls?reverse>
<#list ls as i>
- ${i.name}: ${i.weight}
</#list>

Contains:
---------

<#macro test></#macro>
<#assign x = [1, "2", true, [1,2,3], {"a":1}, test, '1992-02-21'?date('yyyy-MM-dd')]>
True:
${x?seqContains(1.0)?string}
${x?seqContains("2")?string}
${x?seqContains(true)?string}
${x?seqContains('1992-02-21'?date('yyyy-MM-dd'))?string}
${abcSet?seqContains("a")?string}
${abcSet?seqContains("b")?string}
${abcSet?seqContains("c")?string}

False:
${x?seqContains("1")?string}
${x?seqContains(2)?string}
${x?seqContains(false)?string}
${x?seqContains('1992-02-22'?date('yyyy-MM-dd'))?string}
${abcSet?seqContains("A")?string}
${abcSet?seqContains(1)?string}
${abcSet?seqContains(true)?string}

<#assign x = []>
False: ${x?seqContains(1)?string}

Index_of:
---------

<#assign x = [1, "2", true, [1,2,3], {"a":1}, test, '1992-02-21'?date('yyyy-MM-dd')]>
0 = ${x?seqIndexOf(1.0)}
1 = ${x?seqIndexOf("2")}
2 = ${x?seqIndexOf(true)}
6 = ${x?seqIndexOf('1992-02-21'?date('yyyy-MM-dd'))}
0 = ${abcSet?seqIndexOf("a")}
1 = ${abcSet?seqIndexOf("b")}
2 = ${abcSet?seqIndexOf("c")}

-1 = ${x?seqIndexOf("1")}
-1 = ${x?seqIndexOf(2)}
-1 = ${x?seqIndexOf(false)}
-1 = ${x?seqIndexOf('1992-02-22'?date('yyyy-MM-dd'))}
-1 = ${abcSet?seqIndexOf("A")}
-1 = ${abcSet?seqIndexOf(1)}
-1 = ${abcSet?seqIndexOf(true)}

<#assign x = []>
-1 = ${x?seqIndexOf(1)}

Last_index_of:
--------------

<#assign x = [1, "2", true, [1,2,3], {"a":1}, test, 1, '1992-02-21'?date('yyyy-MM-dd')]>
6 = ${x?seqLastIndexOf(1.0)}
1 = ${x?seqLastIndexOf("2")}
2 = ${x?seqLastIndexOf(true)}
7 = ${x?seqLastIndexOf('1992-02-21'?date('yyyy-MM-dd'))}
-1 = ${x?seqLastIndexOf("1")}
0 = ${abcSet?seqLastIndexOf("a")}
1 = ${abcSet?seqLastIndexOf("b")}
2 = ${abcSet?seqLastIndexOf("c")}
-1 = ${abcSet?seqLastIndexOf("A")}

Index_of and last_index_of with starting indices
------------------------------------------------

<#assign names = ["Joe", "Fred", "Joe", "Susan"]>
seq_index_of "Joe":
0 = ${names?seqIndexOf("Joe", -2)}
0 = ${names?seqIndexOf("Joe", -1)}
0 = ${names?seqIndexOf("Joe", 0)}
2 = ${names?seqIndexOf("Joe", 1)}
2 = ${names?seqIndexOf("Joe", 2)}
-1 = ${names?seqIndexOf("Joe", 3)}
-1 = ${names?seqIndexOf("Joe", 4)}
 
seq_last_index_of "Joe":
-1 = ${names?seqLastIndexOf("Joe", -2)}
-1 = ${names?seqLastIndexOf("Joe", -1)}
0 = ${names?seqLastIndexOf("Joe", 0)}
0 = ${names?seqLastIndexOf("Joe", 1)}
2 = ${names?seqLastIndexOf("Joe", 2)}
2 = ${names?seqLastIndexOf("Joe", 3)}
2 = ${names?seqLastIndexOf("Joe", 4)}
 
seq_index_of "Susan":
3 = ${names?seqIndexOf("Susan", -2)}
3 = ${names?seqIndexOf("Susan", -1)}
3 = ${names?seqIndexOf("Susan", 0)}
3 = ${names?seqIndexOf("Susan", 1)}
3 = ${names?seqIndexOf("Susan", 2)}
3 = ${names?seqIndexOf("Susan", 3)}
-1 = ${names?seqIndexOf("Susan", 4)}
 
seq_last_index_of "Susan":
-1 = ${names?seqLastIndexOf("Susan", -2)}
-1 = ${names?seqLastIndexOf("Susan", -1)}
-1 = ${names?seqLastIndexOf("Susan", 0)}
-1 = ${names?seqLastIndexOf("Susan", 1)}
-1 = ${names?seqLastIndexOf("Susan", 2)}
3 = ${names?seqLastIndexOf("Susan", 3)}
3 = ${names?seqLastIndexOf("Susan", 4)}

seq_index_of "a":
0 = ${abcSet?seqIndexOf("a", -2)}
0 = ${abcSet?seqIndexOf("a", -1)}
0 = ${abcSet?seqIndexOf("a", 0)}
-1 = ${abcSet?seqIndexOf("a", 1)}
-1 = ${abcSet?seqIndexOf("a", 2)}
-1 = ${abcSet?seqIndexOf("a", 3)}
-1 = ${abcSet?seqIndexOf("a", 4)}

seq_index_of "b":
1 = ${abcSet?seqIndexOf("b", -2)}
1 = ${abcSet?seqIndexOf("b", -1)}
1 = ${abcSet?seqIndexOf("b", 0)}
1 = ${abcSet?seqIndexOf("b", 1)}
-1 = ${abcSet?seqIndexOf("b", 2)}
-1 = ${abcSet?seqIndexOf("b", 3)}

seq_index_of "c":
2 = ${abcSet?seqIndexOf("c", -2)}
2 = ${abcSet?seqIndexOf("c", -1)}
2 = ${abcSet?seqIndexOf("c", 0)}
2 = ${abcSet?seqIndexOf("c", 1)}
2 = ${abcSet?seqIndexOf("c", 2)}
-1 = ${abcSet?seqIndexOf("c", 3)}
 
seq_last_index_of "a":
-1 = ${abcSet?seqLastIndexOf("a", -2)}
-1 = ${abcSet?seqLastIndexOf("a", -1)}
0 = ${abcSet?seqLastIndexOf("a", 0)}
0 = ${abcSet?seqLastIndexOf("a", 1)}
0 = ${abcSet?seqLastIndexOf("a", 2)}
0 = ${abcSet?seqLastIndexOf("a", 3)}
0 = ${abcSet?seqLastIndexOf("a", 4)}

seq_last_index_of "b":
-1 = ${abcSet?seqLastIndexOf("b", -2)}
-1 = ${abcSet?seqLastIndexOf("b", -1)}
-1 = ${abcSet?seqLastIndexOf("b", 0)}
1 = ${abcSet?seqLastIndexOf("b", 1)}
1 = ${abcSet?seqLastIndexOf("b", 2)}
1 = ${abcSet?seqLastIndexOf("b", 3)}

seq_last_index_of "c":
-1 = ${abcSet?seqLastIndexOf("c", -2)}
-1 = ${abcSet?seqLastIndexOf("c", -1)}
-1 = ${abcSet?seqLastIndexOf("c", 0)}
-1 = ${abcSet?seqLastIndexOf("c", 1)}
2 = ${abcSet?seqLastIndexOf("c", 2)}
2 = ${abcSet?seqLastIndexOf("c", 3)}

Sequence builtins ignoring nulls
--------------------------------

true = ${listWithNull?seqContains('c')?string}
2 = ${listWithNull?seqIndexOf('c')}
0 = ${listWithNull?seqLastIndexOf('a')}

These should throw exception, but for BC they don't:
false = ${listWithNull?seqContains(noSuchVar)?string}
-1 = ${listWithNull?seqIndexOf(noSuchVar)}
-1 = ${listWithNull?seqLastIndexOf(noSuchVar)}

Sequence built-ins failing on date-type mismatch
------------------------------------------------

<#assign x = ['1992-02-21'?date('yyyy-MM-dd'), 'foo']>
<@assertEquals actual=x?seqIndexOf('foo') expected=1 />
<@assertEquals actual=x?seqIndexOf('1992-02-21'?date('yyyy-MM-dd')) expected=0 />
<@assertFails message="dates of different types">
  0 = ${x?seqIndexOf('1992-02-21 00:00:00'?dateTime('yyyy-MM-dd HH:mm:ss'))}
</@>

Chunk
-----

<#assign ls = ['a', 'b', 'c', 'd', 'e', 'f', 'g']>
<#list ['NULL', '-'] as fill>
  <#list [1, 2, 3, 4, 5, 10] as columns>
    <@printTable ls columns=columns fill=fill />
  </#list>
</#list>
<@printTable [1, 2, 3, 4, 5, 6, 7, 8, 9] columns=3 fill='NULL' />
<@printTable [1, 2, 3, 4, 5, 6, 7, 8, 9] columns=3 fill='-' />
<@printTable [1] columns=3 fill='NULL' />
<@printTable [1] columns=3 fill='-' />
<@printTable [] columns=3 fill='NULL' />
<@printTable [] columns=3 fill='-' />

<#macro printTable ls{positional} columns fill>
  columns = ${columns}, fill = ${fill}:<#lt>
  <#if fill=='NULL'>
    <#local rows = ls?chunk(columns)>
  <#else>
    <#local rows = ls?chunk(columns, fill)>
  </#if>
  Rows: ${rows?size}
  <#list rows as row>
    <#list row as i>${i} </#list>  <-- Columns: ${row?size}
  </#list>
  
</#macro>


Join
----

<#assign xs = [1, "two", "three", 4]>
- ${xs?join(", ")}
- ${[]?join(", ")}
- ${xs?join(", ", "(empty)", ".")}
- ${[]?join(", ", "(empty)", ".")}
- ${listWithNull?join(", ")}
- ${listWithNull?join(", ", "(empty)")}
- ${listWithNull?join(", ", "(empty)", ".")}
- ${listWithNullsOnly?join(", ")}
- ${listWithNullsOnly?join(", ", "(empty)")}
- ${listWithNullsOnly?join(", ", "(empty)", ".")}
- ${abcSet?join(", ", "(empty)", ".")}
- ${abcCollection?join(", ", "(empty)", ".")}
<@assertFails message="index 1">${['a', [], 'c']?join(", ", "(empty)", ".")}</@>

Misc
----

First of set 1: ${abcSet?first}
First of set 2: ${abcSetNonSeq?first}