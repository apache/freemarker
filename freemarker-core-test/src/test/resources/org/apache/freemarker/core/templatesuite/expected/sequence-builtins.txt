/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
Sorting strings:
----------------

String order:
- aardvark
- Barbara
- beetroot
- whale
- zeppelin

First: aardvark
Last: zeppelin
Size 5

Numerical order:
- -324
- -34
- 0
- 0.1
- 0.11
- 1
- 5
- 111
- 123
- 543

First: -324
Last: 543
Size 10

Date/time order:
- 06:05
- 08:05
- 08:15
- 18:00

Boolean order:
- false
- false
- true
- true


Sorting hashes:
---------------

Order by name:
- aardvark: 30
- Barbara: 53
- beetroot: 0.3
- whale: 2000
- zeppelin: -200

Order by weight:
- zeppelin: -200
- beetroot: 0.3
- aardvark: 30
- Barbara: 53
- whale: 2000

Order by a.x.v:
- aqweqw
- dfgdf
- qweqw
- utyu

Order by a.y, which is a date:
- 1998-02-20
- 1999-01-20
- 1999-04-19
- 1999-04-20

Reverse:
--------

Order by weight desc:
- whale: 2000
- Barbara: 53
- aardvark: 30
- beetroot: 0.3
- zeppelin: -200

Order by weight desc desc:
- zeppelin: -200
- beetroot: 0.3
- aardvark: 30
- Barbara: 53
- whale: 2000

Order by weight desc desc desc:
- whale: 2000
- Barbara: 53
- aardvark: 30
- beetroot: 0.3
- zeppelin: -200

Contains:
---------

True:
true
true
true
true
true
true
true

False:
false
false
false
false
false
false
false

False: false

Index_of:
---------

0 = 0
1 = 1
2 = 2
6 = 6
0 = 0
1 = 1
2 = 2

-1 = -1
-1 = -1
-1 = -1
-1 = -1
-1 = -1
-1 = -1
-1 = -1

-1 = -1

Last_index_of:
--------------

6 = 6
1 = 1
2 = 2
7 = 7
-1 = -1
0 = 0
1 = 1
2 = 2
-1 = -1

Index_of and last_index_of with starting indices
------------------------------------------------

seq_index_of "Joe":
0 = 0
0 = 0
0 = 0
2 = 2
2 = 2
-1 = -1
-1 = -1
 
seq_last_index_of "Joe":
-1 = -1
-1 = -1
0 = 0
0 = 0
2 = 2
2 = 2
2 = 2
 
seq_index_of "Susan":
3 = 3
3 = 3
3 = 3
3 = 3
3 = 3
3 = 3
-1 = -1
 
seq_last_index_of "Susan":
-1 = -1
-1 = -1
-1 = -1
-1 = -1
-1 = -1
3 = 3
3 = 3

seq_index_of "a":
0 = 0
0 = 0
0 = 0
-1 = -1
-1 = -1
-1 = -1
-1 = -1

seq_index_of "b":
1 = 1
1 = 1
1 = 1
1 = 1
-1 = -1
-1 = -1

seq_index_of "c":
2 = 2
2 = 2
2 = 2
2 = 2
2 = 2
-1 = -1
 
seq_last_index_of "a":
-1 = -1
-1 = -1
0 = 0
0 = 0
0 = 0
0 = 0
0 = 0

seq_last_index_of "b":
-1 = -1
-1 = -1
-1 = -1
1 = 1
1 = 1
1 = 1

seq_last_index_of "c":
-1 = -1
-1 = -1
-1 = -1
-1 = -1
2 = 2
2 = 2

Sequence builtins ignoring nulls
--------------------------------

true = true
2 = 2
0 = 0

These should throw exception, but for BC they don't:
false = false
-1 = -1
-1 = -1

Sequence built-ins failing on date-type mismatch
------------------------------------------------


Chunk
-----

columns = 1, fill = NULL:
  Rows: 7
    a   <-- Columns: 1
    b   <-- Columns: 1
    c   <-- Columns: 1
    d   <-- Columns: 1
    e   <-- Columns: 1
    f   <-- Columns: 1
    g   <-- Columns: 1
  
columns = 2, fill = NULL:
  Rows: 4
    a b   <-- Columns: 2
    c d   <-- Columns: 2
    e f   <-- Columns: 2
    g   <-- Columns: 1
  
columns = 3, fill = NULL:
  Rows: 3
    a b c   <-- Columns: 3
    d e f   <-- Columns: 3
    g   <-- Columns: 1
  
columns = 4, fill = NULL:
  Rows: 2
    a b c d   <-- Columns: 4
    e f g   <-- Columns: 3
  
columns = 5, fill = NULL:
  Rows: 2
    a b c d e   <-- Columns: 5
    f g   <-- Columns: 2
  
columns = 10, fill = NULL:
  Rows: 1
    a b c d e f g   <-- Columns: 7
  
columns = 1, fill = -:
  Rows: 7
    a   <-- Columns: 1
    b   <-- Columns: 1
    c   <-- Columns: 1
    d   <-- Columns: 1
    e   <-- Columns: 1
    f   <-- Columns: 1
    g   <-- Columns: 1
  
columns = 2, fill = -:
  Rows: 4
    a b   <-- Columns: 2
    c d   <-- Columns: 2
    e f   <-- Columns: 2
    g -   <-- Columns: 2
  
columns = 3, fill = -:
  Rows: 3
    a b c   <-- Columns: 3
    d e f   <-- Columns: 3
    g - -   <-- Columns: 3
  
columns = 4, fill = -:
  Rows: 2
    a b c d   <-- Columns: 4
    e f g -   <-- Columns: 4
  
columns = 5, fill = -:
  Rows: 2
    a b c d e   <-- Columns: 5
    f g - - -   <-- Columns: 5
  
columns = 10, fill = -:
  Rows: 1
    a b c d e f g - - -   <-- Columns: 10
  
columns = 3, fill = NULL:
  Rows: 3
    1 2 3   <-- Columns: 3
    4 5 6   <-- Columns: 3
    7 8 9   <-- Columns: 3
  
columns = 3, fill = -:
  Rows: 3
    1 2 3   <-- Columns: 3
    4 5 6   <-- Columns: 3
    7 8 9   <-- Columns: 3
  
columns = 3, fill = NULL:
  Rows: 1
    1   <-- Columns: 1
  
columns = 3, fill = -:
  Rows: 1
    1 - -   <-- Columns: 3
  
columns = 3, fill = NULL:
  Rows: 0
  
columns = 3, fill = -:
  Rows: 0
  



Join
----

- 1, two, three, 4
- 
- 1, two, three, 4.
- (empty)
- a, c
- a, c
- a, c.
- 
- (empty)
- (empty)
- a, b, c.
- a, b, c.


Misc
----

First of set 1: a
First of set 2: a